### PR TITLE
Add ConvRelu shape/input info in Caffe2ModelLoader.cpp

### PR DESCRIPTION
**Pull request by zrphercule**
*Description*:
The ConvRelu is same to Conv or Int8Conv, so just use it directly.

### CODE REVIEW

**Comment by beicy** [comment]
In this case, ConvRelu will be treated as "Conv". I think we should split it into "conv" + "Rule" ?

**Comment by zrphercule** [comment]
> In this case, ConvRelu will be treated as "Conv". I think we should split it into "conv" + "Rule" ?

But I noticed Int8ConvRelu is also here already, and it is correct

**Comment by beicy** [comment]
> > In this case, ConvRelu will be treated as "Conv". I think we should split it into "conv" + "Rule" ?
> 
> But I noticed Int8ConvRelu is also here already, and it is correct

Right. But I mean under this switch, we have different flows to handle different op. There are a few branches " if (typeName == "Conv")" or " if (typeName != "Conv") ".

**Comment by jfix71** [comment]
From what I can see we aren't actually creating a relu for `Int8ConvRelu `, no? This seems broken if so.

**Comment by beicy** [comment]
> From what I can see we aren't actually creating a relu for `Int8ConvRelu `, no? This seems broken if so.

No. After some discussion, we found for us, adding "Int8Relu" here is kind of useless.

**Comment by rdzhabarov** [comment]
As we've already have proper {scale,offset} from the the ConvRelu node (covering non negative range).
Adding additional ReLU seems to be redundant as proper range is already covered by setting proper quantization params for Conv.

**Comment by rdzhabarov** [comment]
Please, run clang-format @zrphercule

**Comment by jfix71** [comment]
>As we've already have proper {scale,offset} from the the ConvRelu node (covering non negative range). Adding additional ReLU seems to be redundant as proper range is already covered by setting proper quantization params for Conv.

Ok, that seems like something we should at least note in the importer logic here. I think adding an assert or check that the quantized range is only positive would be a good idea.

**Comment by rdzhabarov** [comment]
> I think adding an assert or check that the quantized range is only positive would be a good idea.

Totally.

**Comment by rdzhabarov** [comment]
@zrphercule good to go after addressing Jordan's concern.

File: lib/Importer/Caffe2ModelLoader.cpp [hunk 1]
```diff
@@ -237,7 +237,7 @@ llvm::Error Caffe2ModelLoader::loadOperator(const caffe2::OperatorDef &op) {
   }
   const std::string &opName = loadOperatorName(op);
 
-  if (typeName == "Conv" || typeName == "Int8Conv" ||
+  if (typeName == "Conv" || typeName == "ConvRelu" || typeName == "Int8Conv" ||
```
Comment (nadavrot) [inline lib/Importer/Caffe2ModelLoader.cpp]: @zrphercule Is this correct? Where are you constructing the Relu in the code below?

Comment (jfix71) [reply lib/Importer/Caffe2ModelLoader.cpp]: Yeah, I think we need to create a ReLU for the float version. It only makes sense to skip the ReLU for the quantized version assuming that the Conv has non-negative quantization range.

**Comment by zrphercule** [comment]
@rdzhabarov I wonder if this part is only for shape info, or it is for building the graph? If it includes build the graph, then I guess we cannot simply change it this way, since we cannot ignore the Relu in a normal ConvRelu.

**Comment by zrphercule** [comment]
Hi all, I would like to add a Relu node here, so no need for more review for now. Will tag you guys once I finished hahah! Thanks!

**Comment by rdzhabarov** [comment]
>  I wonder if this part is only for shape info, or it is for building the graph?

This is actual graph building.

I thought originally that this is for quantized ConvRelu, but appears to be it's actually the other way around. It's only for fp32. So, as we discussed let's verify that we operate on fp32 and do immediate lowering (Conv followed by Relu) in the loader.

Alternative solution would be to unfuse that operator on the ONNXIFI level which can be done at a later point. Loading ConvRelu by standalone Glow seems useful as well.

**Comment by zrphercule** [comment]
@rdzhabarov Wonder if this is what we need?

File: lib/Importer/Caffe2ModelLoader.cpp [hunk 2]
```diff
@@ -355,7 +355,10 @@ llvm::Error Caffe2ModelLoader::loadOperator(const caffe2::OperatorDef &op) {
 
     Node *node = G_.createConv(opName, finalIn, filter, bias, outTy, kernels,
                                strides, pads, group);
-
+    if (typeName == "ConvRelu") {
```
Comment (rdzhabarov) [inline lib/Importer/Caffe2ModelLoader.cpp]: You need to add a test to Caffe2ImporterTest to verify ConvRelu.

File: lib/Importer/Caffe2ModelLoader.cpp [hunk 3]
```diff
@@ -355,7 +355,10 @@ llvm::Error Caffe2ModelLoader::loadOperator(const caffe2::OperatorDef &op) {
...
+      NodeValue relu_in = node->getNthResult(0);
+      node = G_.createRELU(opName, relu_in);
```
Comment (rdzhabarov) [inline lib/Importer/Caffe2ModelLoader.cpp]: node = G_.createRELU(opName, node) should be enough. (I'd add opName + ".relu")

File: tests/unittests/Caffe2ImporterTest.cpp [hunk 4]
```diff
@@ -67,6 +67,43 @@ TEST(caffe2, importConv) {
...
+    Caffe2ModelLoader caffe2LD(NetDescFilename, NetWeightFilename,
+                               {"gpu_0/data_0"}, {&data.getType()}, *F);
+    output = EXIT_ON_ERR(caffe2LD.getSingleOutput());
+
+    bindings.allocate(mod.getPlaceholders());
+    updateInputPlaceholdersByName(bindings, &mod, {"gpu_0/data_0"}, {&data});
+  }
+
+  auto res = bindings.get(output);
+  EE.compile(CompilationMode::Infer, F);
```
Comment (rdzhabarov) [inline tests/unittests/Caffe2ImporterTest.cpp]: for importer tests the goal is to verify structure, can you check that there are Conv, Relu nodes here.

**Comment by rdzhabarov** [comment]
Test comment, the rest LGTM

**Comment by zrphercule** [comment]
A reminder to myself: dont forget to add comments.

**Pull request merged**